crates/win-api-wrappers/src/wts.rs

```diff
@@ -143,40 +141,45 @@ pub struct WTSSessionInfo {
 impl WTSSessionInfo {
     /// # Safety
     ///
-    /// - The `pWinStationName` pointer needs to be valid for reads up until and including the next `\0`.
+    /// - The `wts_session.pWinStationName` pointer must be valid for reads up until and including the next `\0`.
     unsafe fn new(wts_session: &WTS_SESSION_INFOW) -> anyhow::Result<Self> {
+        // SAFETY: Same preconditions as the current function.
+        let win_station_name = unsafe { wts_session.pWinStationName.to_string()? };
+
         Ok(Self {
             session_id: wts_session.SessionId,
-            win_station_name: unsafe { wts_session.pWinStationName.to_string()? },
+            win_station_name,
             state: wts_session.State.try_into()?,
         })
     }
 }
 
 pub fn get_sessions() -> anyhow::Result<Vec<WTSSessionInfo>> {
-    let mut sessions_ptr: *mut WTS_SESSION_INFOW = std::ptr::null_mut();
-    let mut len = 0u32;
+    let mut sessions: *mut WTS_SESSION_INFOW = core::ptr::null_mut();
+    let mut count = 0u32;
 
-    // SAFETY: All buffers are valid.
+    // SAFETY: FFI call with no outstanding precondition.
     unsafe {
-        WTSEnumerateSessionsW(WTS_CURRENT_SERVER_HANDLE, 0, 1, &mut sessions_ptr, &mut len)?;
+        WTSEnumerateSessionsW(WTS_CURRENT_SERVER_HANDLE, 0, 1, &mut sessions, &mut count)?;
     };
 
-    // SAFETY: `sessions_ptr` is always a valid pointer on success.
-    // `sessions_ptr` must be freed by `WTSFreeMemory`.
-    let sessions_ptr = unsafe { WTSMemory::from_raw(sessions_ptr.cast()) };
-
-    // SAFETY: Verify that all the safety preconditions of from_raw_parts are uphold: https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety
-    let sessions_slice: &[WTS_SESSION_INFOW] =
-        unsafe { std::slice::from_raw_parts(sessions_ptr.0 as *mut u8 as *mut WTS_SESSION_INFOW, len as usize) };
-
-    let mut sessions = Vec::<WTSSessionInfo>::with_capacity(sessions_slice.len());
-    for session in sessions_slice {
-        // SAFETY: `session` is valid
-        unsafe {
-            sessions.push(WTSSessionInfo::new(session)?);
-        }
-    }
+    // SAFETY: The pointer placed into `sessions` by `WTSEnumerateSessionsW` must be freed by `WTSFreeMemory`.
+    let sessions = unsafe { WtsMemory::from_raw(sessions) };
+
+    // SAFETY:
+    // - `WTSEnumerateSessionsW` returns a pointer valid for `count` reads of `WTS_SESSION_INFOW`.
+    // - `WTSEnumerateSessionsW` is also returning a pointer properly aligned.
+    // - We ensure the memory referenced by the slice is not mutated by shadowing the variable.
+    // - There are never so many sessions that `count * mem::size_of::<WTS_SESSION_INFOW>()` overflows `isize`.
+    let sessions = unsafe { sessions.cast_slice(count as usize) };
```

Comment:
`cast_slice` is tying the lifetime of the slice to the lifetime of the WtsMemory object, preventing use-after-free bugs.